File: 3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        nums=sorted(nums)
        for k in range(len(nums)-1):
            if k>0 and nums[k]==nums[k-1]:
                continue
            i=k+1
            j=len(nums)-1
            while i<j:
                if nums[i]+nums[j]+nums[k]>0 :
                    j-=1
                elif nums[i]+nums[j]+nums[k]<0 :
                    i+=1
                else:
                    ans.append([nums[i],nums[j],nums[k]])
                    i+=1
                    while i<j and nums[i]==nums[i-1]:
                        i+=1
                    j-=1
                    while i<j and nums[j]==nums[j+1]:
                        j-=1
        return ans

File: 3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>>ans;
        for(int i=0;i<nums.size()-2;i++){
           if(i>0&&nums[i]==nums[i-1])continue;
           int l=i+1;
           int h=nums.size()-1;
           while(l<h){
                if(nums[i]+nums[l]+nums[h]==0){
                    ans.push_back({nums[i],nums[l],nums[h]});
                    l++;
                    while(nums[l]==nums[l-1]&&l<h)
                        l++;
                    h--;
                }else if(nums[i]+nums[l]+nums[h]>0){
                    h--;
                }else{
                    l++;
                }
           }
        }
        return ans;
    }
};